be/src/exec/sorting/merge_path.h

```diff
@@ -0,0 +1,434 @@
+     */
+    static void _eval_diagnoal_intersection(const SortDescs& descs, const InputSegment& left, const InputSegment& right,
+                                            const size_t d_size, const size_t parallel_idx,
+                                            const size_t degree_of_parallelism, size_t& l_start, size_t& r_start);
+
+    /**
+     * Check if the point (left[li], right[ri]) is intersection point.
+     * Assuming M matrix is a matrix conprising of only boolean value
+     *      if left[i] > right[j], then M[i, j] = true
+     *      if left[i] <= right[j], then M[i, j] = false
+     * For edge cases (i or j beyond the matrix), think about the merge path, with left in the vertical direction 
+     * and right in the horizontal direction. Merge path goes from top left to bottom right, the positions below the merge path should be true, 
+     * and otherwise should be false.
+     *
+     * And the following 4 points will be checked
+     *      1. (left[li-1], right[ri-1])
+     *      2. (left[li-1], right[ri])
+     *      3. (left[li], right[ri-1])
+     *      4. (left[li], right[ri])
+     *
+     *
+     * @param desc Ordering description.
+     * @param left Left side input.
+     * @param li Current index in left.
+     * @param right Right side input.
+     * @param ri Current index in right.
+     * @param has_true Output param, true if one of the above four points is true.
+     * @param has_false Ouput param, true if one of the above four points is false.
+     */
+    static void _is_intersection(const SortDescs& descs, const InputSegment& left, const size_t li,
+                                 const InputSegment& right, const size_t ri, bool& has_true, bool& has_false);
+
+    /**
+     * Do merge along the merge path.
+     *
+     * @param desc Ordering description.
+     * @param left Left side input.
+     * @param li Current index in left.
+     * @param right Right side input.
+     * @param ri Current index in right.
+     * @param dest Output used to store the merged data for this parallelism.
+     * @param start_di Start index of the whole merge result, NOT only the above dest.
+     * @param length The step for this parallelism can forward along merge path.
+     */
+    static void _do_merge_along_merge_path(const SortDescs& descs, const InputSegment& left, size_t& li,
+                                           const InputSegment& right, size_t& ri, OutputSegment& dest, size_t start_di,
+                                           const size_t length);
+};
+
+class MergePathCascadeMerger;
+
+/**
+ * LeafNode will use this provider to get data.
+ *
+ * @param only_check If true, only check if it has more data with fetching chunk
+ * @param chunk Output chunk if any
+ * @param eos Set to true if this provider reaches eos
+ * @return: Return true if it has new data
+ */
+using MergePathChunkProvider = std::function<bool(bool only_check, ChunkPtr* chunk, bool* eos)>;
+
+namespace detail {
```

Comment:
should we give it a more meaningful name? starrocks::detail does not seem to be exclusive to merge path module

Reply:
Updated, please check again if it is appropriate